sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/Constraints/SymbolicValueConstraints.cs

```diff
@@ -0,0 +1,115 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using SonarAnalyzer.Common;
+
+namespace SonarAnalyzer.Helpers.FlowAnalysis.Common
+{
+    public class SymbolicValueConstraints
+    {
+        private readonly Dictionary<Type, SymbolicValueConstraint> constraints
+            = new Dictionary<Type, SymbolicValueConstraint>();
+
+        private readonly int hashCode;
+
+        public SymbolicValueConstraints(SymbolicValueConstraint constraint)
+        {
+            SetConstraint(constraint, this.constraints);
+            hashCode = ComputeHashcode();
+        }
+
+        private SymbolicValueConstraints(Dictionary<Type, SymbolicValueConstraint> constraints)
+        {
+            this.constraints = constraints;
+            hashCode = ComputeHashcode();
+        }
+
+        internal IEnumerable<SymbolicValueConstraint> GetConstraints() => constraints.Values;
+
+        internal virtual SymbolicValueConstraints SetConstraint(SymbolicValueConstraint constraint)
+        {
+            var constraintsCopy = new Dictionary<Type, SymbolicValueConstraint>(constraints);
+            SetConstraint(constraint, constraintsCopy);
+
+            return new SymbolicValueConstraints(constraintsCopy);
+        }
+
+        private static void SetConstraint(SymbolicValueConstraint constraint,
+            Dictionary<Type, SymbolicValueConstraint> constraints)
+        {
+            constraints[constraint.GetType()] = constraint;
+
+            if (constraint is BoolConstraint ||
+                constraint is DisposableConstraint)
+            {
+                constraints[typeof(ObjectConstraint)] = ObjectConstraint.NotNull;
+            }
+        }
+
+        internal T GetConstraint<T>()
+            where T : SymbolicValueConstraint
+        {
+            SymbolicValueConstraint constraint = null;
+            return constraints.TryGetValue(typeof(T), out constraint)
+                ? (T)constraint
+                : null;
+        }
+
+        internal bool HasConstraint(SymbolicValueConstraint constraint)
+        {
+            SymbolicValueConstraint storedConstraint;
+            return constraints.TryGetValue(constraint.GetType(), out storedConstraint) &&
+                   storedConstraint == constraint;
+        }
+
+        private int ComputeHashcode()
+        {
+            int hash = 17 * constraints.Count;
+
+            foreach (var item in constraints)
+            {
+                hash = hash * 23 + item.Value.GetHashCode();
+            }
+
+            return hash;
+        }
+
+        public override int GetHashCode()
+        {
+            return hashCode;
+        }
+
+        public override bool Equals(object obj)
+        {
+            var other = obj as SymbolicValueConstraints;
+            return other != null
+                ? DictionaryHelper.DictionaryEquals(constraints, other.constraints)
+                : false;
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the unnecessary Boolean literal(s). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)